### DISCUSSION TITLE
Question about answer_fn() and question_numeric()

### DISCUSSION THREAD

**Discussion by databeej2w** [rstudio/learnr#797] [Q&A] [answered]
I am creating a simple learnr exercise with a numerical answer using question_numeric(). As the parameter in the questions is randomly generated and the answer depends on the randomly generated parameter, I need to use the answer_fn() function within the question_numeric(). Specifically, answers within a certain range are all acceptable and the range depends on the parameter. I cannot solve this by specifying tolerance since the acceptable range varies with the parameter. Hence, I need to pass the value of this parameter to the function defined within answer_fn().

The example given in the Help page for question_numeric() below
https://rstudio.github.io/learnr/reference/question_numeric.html

```
question_numeric(
  "Can you think of an even number?",
  answer_fn(function(value) {
    if (value %% 2 == 0) {
      correct("even")
    } else if (value %% 2 == 1) {
      incorrect("odd")
    }
  }, label = "Is the number even?"),
  step = 1
)
```
the function defined within answer_fn() only takes one input (the submitted answer). Is there a way to pass the parameter to the function defined within answer_fn()? Thank you!

More specifically, I want to create a question template that asks students to find the probability P(Z < z) for Z ~ N(0,1) using a normal probability table, where the value of z is randomly generated. As the normal table only give values for pnorm(z) for z up to 2 decimal places, if z has 3 or more decimal places, e.g., z = 2.343, then the learnr question should accept any values between pnorm(2.34) and pnorm(2.35) as the correct answer. Thus I need to write an answer_fn() function that takes z as an argument. Does anyone know how to do that?

Below is the R codes I made, which doesn't work.

```{r NormLowerTail, echo=FALSE}
z = round(runif(1, min = -3.49, max = 3.49),3)
question_numeric(
  paste("Find the probability $P(Z < ", z, ")$ if $Z$ has a standard normal distribution N(0,1)? Keep **4 decimal places*."),
  answer_fn(function(value,z) {
    z_floor = ifelse(z >= round(z,2), round(z,2), round(z,2) - 0.01)
    z_ceiling = ifelse(z <= round(z,2), round(z,2) , round(z,2) + 0.01)
    if (value >= pnorm(z_floor) & value <= pnorm(z_ceiling)) {
      correct("Correct!")
    }
  }),
  # answer(round(pnorm(z),4), correct = TRUE),
  allow_retry = TRUE,
  min = 0,
  max = 1,
  step = 0.0001
)
```

Best Regards,
Yibi

**Comment by databeej2w** [comment]
I updated my question above. I describe the function I intended to create in greater detail.

**Comment by gadenbuie** [comment, marked as answer]
The core problem you're encountering is that between rendering your tutorial's `.Rmd` file and running the Shiny app that the student sees, the question created with `question_numeric()` is written into static data. While the `answer_fn()` approach lets you define some code that is used to determine whether or not the student's answer is correct, technically in the interim the function body is written out as a character string.

In other words, when the question is shown to the user, it has to be completely isolated from the code that was used to create the question. This is why `answer_fn()` needs a function that only takes `value` as an argument.

The way to get around this is a little tricky. Basically you create the answer function outside of the `question_numeric()` call. This function takes `value` and `z`, but you need the default value of `z` to be the literal value that you randomly selected, not a variable that resolves to that value (since the variable won't be available in the environment when the question is running for the student).

There's a function in base R that shows you the arguments of a function, called `formals()`. You can even use `formals()` to update the function signature of an existing function. So we can use `formals(answer_z)$z <- z` to assign the _value of_ the randomly selected `z` to the default value of the `z` argument.

Here's the final code that creates the question you want:

```r
# Pick a random z value
z = round(runif(1, min = -3.49, max = 3.49), 3)

# A grading function give the student's `value` and an *arbitrary* `z`
answer_z <- function(value, z) {
  z_floor = ifelse(z >= round(z,2), round(z,2), round(z,2) - 0.01)
  z_ceiling = ifelse(z <= round(z,2), round(z,2) , round(z,2) + 0.01)
  if (value >= pnorm(z_floor) & value <= pnorm(z_ceiling)) {
    correct()
  }
}

# Assign the literal value of our randomly selected `z` to be used as
# the default value of the grading function's `z` argument
formals(answer_z)$z <- z

question_numeric(
  paste0(
    "Find the probability $P(Z < ", z, ")$ if ",
    "$Z$ has a standard normal distribution N(0,1)? ",
    "Keep *4 decimal places*. ",
    "(The answer is ~", signif(pnorm(z), 6), ".)"
  ),
  answer_fn(answer_z),
  # answer(round(pnorm(z),4), correct = TRUE),
  allow_retry = TRUE,
  min = 0,
  max = 1,
  step = 0.0001
)
```

**Reply by databeej2w** [reply]
Thank you so much! That solves my problem.
